### PR TITLE
Add serializability check for interface payloads in Haskell

### PR DESCRIPTION
**Pull request by cocreature**
changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by cocreature** [comment]
Actually this doesn’t work. It breaks the serializability check for contract ids which should be serializable.

**Comment by cocreature** [comment]
confused myself very successfully, trying to cleanup first https://github.com/digital-asset/daml/pull/12565

File: compiler/damlc/tests/daml-test-files/InterfaceSerializabilityPayload.daml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+-- @SINCE-LF-FEATURE DAML_INTERFACE
+-- @ERROR expected serializable type
+module InterfaceSerializabilityPayload where
+
+-- Test that the interface serializability payload itself is not serializable.
+interface Gettable where
+  nonconsuming choice Get : Gettable
```
Comment (akrmn) [inline compiler/damlc/tests/daml-test-files/InterfaceSerializabilityPayload.daml:7]: 🤯

**Comment by cocreature** [comment]
Alright I think this works now and is reviewable.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs [hunk 2]
```diff
@@ -29,26 +30,50 @@ import DA.Daml.LF.Ast.Optics (_PRSelfModule, dataConsType)
...
+                | Right tconName <- matching (_PRSelfModule $ modName currentModule) con
+                -> if tconName `HS.member` modInterfaces currentModule
+                     then noConditions
+                     else go typ
+              (Left err, _) -> error $ showString "Serializability.serializabilityConditionsDataTyp: " $ show err
+          | TCon con <- typ
+          , Just currentModule <- mbCurrentModule
+          , Right tconName <- matching (_PRSelfModule $ modName currentModule) con
+          , tconName `HS.member` modInterfaces currentModule -> noConditions
+          | otherwise -> go typ
```
Comment (sofiafaro-da) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs:76]: Just a suggestion, I would try to factor out an `isInterface` here that just determines if typ refers to an interface type constructor.

```suggestion
      TContractId typ ->
          if isInterface world0 mbCurrentModule typ
            then noConditions
            else go typ
```

**Review by sofiafaro-da — APPROVED** [review]
Nice work, thank you!

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs (hunk 2 above)
Comment (cocreature) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs]: good idea done

**Pull request merged**